File: keygen.py
#!/usr/bin/env python3
"""
Ultra-secure license key generator for yurucode
5x5 format with HMAC-SHA256 signatures
"""

import os
import sys
import time
import hmac
import hashlib
import secrets
from pathlib import Path

# Load server secrets from .env.secret file
def load_env_secrets():
    env_file = Path(__file__).parent / 'yurucode.com' / '.env.secret'
    if env_file.exists():
        with open(env_file, 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, value = line.split('=', 1)
                    os.environ[key.strip()] = value.strip()

# Load environment variables
load_env_secrets()

# Server secrets (must match PHP)
MASTER_SECRET = os.getenv('YURUCODE_MASTER_SECRET', '7f3a9c2e8b4d6f1a5c7e9b3d5f8a2c4e6b8d1f3a5c7e9b2d4f6a8c1e3b5d7f9a2c')
VALIDATION_SECRET = os.getenv('YURUCODE_VALIDATION_SECRET', '9e2b4f7a1c5d8e3b6f9a2c5e8b1d4f7a3c6e9b2d5f8a1c4e7b9d3f6a8c2e5b7d')
TIMESTAMP_SECRET = os.getenv('YURUCODE_TIMESTAMP_SECRET', '4a7c9e2b5d8f1a3c6e9b2d5f8a1c4e7b9d3f6a8c2e5b7d9f2a4c6e8b1d3f5a7c')

# License configuration
LICENSE_ALPHABET = '23456789ABCDEFGHJKLMNPQRSTUVWXYZ'  # 32 chars (no 0,1,I,O)

def encode_segment(num: int) -> str:
    """Encode number to 5-char segment"""
    result = []
    base = len(LICENSE_ALPHABET)
    for _ in range(5):
        result.append(LICENSE_ALPHABET[num % base])
        num //= base
    return ''.join(reversed(result))

def decode_segment(s: str) -> int:
    """Decode 5-char segment to number"""
    num = 0
    base = len(LICENSE_ALPHABET)
    for char in s.upper():
        idx = LICENSE_ALPHABET.find(char)
        if idx == -1:
            raise ValueError(f"Invalid character: {char}")
        num = num * base + idx
    return num

def create_signature(data: str, timestamp: int) -> bytes:
    """Create HMAC signature"""
    msg = f"{data}|{timestamp}|{VALIDATION_SECRET}"
    
    # Multi-layer HMAC
    h1 = hmac.new(MASTER_SECRET.encode(), msg.encode(), hashlib.sha256).digest()
    h2 = hmac.new(TIMESTAMP_SECRET.encode(), h1 + str(timestamp).encode(), hashlib.sha256).digest()
    
    return hashlib.sha256(h1 + h2).digest()

def generate_license_key() -> str:
    """Generate a 5x5 format license key"""
    # Current timestamp
    timestamp = int(time.time())
    
    # Generate 3 random segments
    seg1 = encode_segment(secrets.randbelow(32**5))
    seg2 = encode_segment(secrets.randbelow(32**5))
    seg3 = encode_segment(timestamp % (32**5))  # Embed timestamp in seg3
    
    # Create signature from first 3 segments
    data = f"{seg1}-{seg2}-{seg3}"
    sig = create_signature(data, timestamp)
    
    # Convert signature to 2 segments
    sig_num1 = int.from_bytes(sig[:4], 'big') % (32**5)
    sig_num2 = int.from_bytes(sig[4:8], 'big') % (32**5)
    
    seg4 = encode_segment(sig_num1)
    seg5 = encode_segment(sig_num2)
    
    return f"{seg1}-{seg2}-{seg3}-{seg4}-{seg5}"

def validate_license_key(key: str) -> dict:
    """Validate a license key"""
    try:
        # Check format
        parts = key.upper().split('-')
        if len(parts) != 5 or any(len(p) != 5 for p in parts):
            return {'valid': False, 'error': 'Invalid format'}
        
        # Check characters
        for part in parts:
            for char in part:
                if char not in LICENSE_ALPHABET:
                    return {'valid': False, 'error': f'Invalid character: {char}'}
        
        # Decode timestamp from seg3
        timestamp = decode_segment(parts[2])
        
        # Basic timestamp validation
        current = int(time.time())
        # The timestamp in seg3 is modulo 33^5, so we need to find the actual timestamp
        # Look for timestamps in reasonable range (last 2 years to future)
        base_timestamp = timestamp
        possible_timestamps = []
        
        max_seg_value = 32**5
        # Start from 0 and find all possible timestamps
        i = 0
        while True:
            test_timestamp = base_timestamp + (i * max_seg_value)
            if test_timestamp > current + 86400:  # More than 1 day future
                break
            if test_timestamp > current - (365 * 86400 * 2):  # Within 2 years
                possible_timestamps.append(test_timestamp)
            i += 1
            if i > 1000:  # Safety limit
                break
        
        # Try each possible timestamp
        data = f"{parts[0]}-{parts[1]}-{parts[2]}"
        for ts in reversed(possible_timestamps):  # Try newest first
            sig = create_signature(data, ts)
            
            sig_num1 = int.from_bytes(sig[:4], 'big') % (32**5)
            sig_num2 = int.from_bytes(sig[4:8], 'big') % (32**5)
            
            expected_seg4 = encode_segment(sig_num1)
            expected_seg5 = encode_segment(sig_num2)
            
            if parts[3] == expected_seg4 and parts[4] == expected_seg5:
                from datetime import datetime
                return {
                    'valid': True,
                    'timestamp': ts,
                    'issued': datetime.fromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S'),
                    'age_days': (current - ts) // 86400
                }
        
        return {'valid': False, 'error': 'Invalid signature'}
        
    except Exception as e:
        return {'valid': False, 'error': str(e)}

def main():
    """CLI interface"""
    action = sys.argv[1] if len(sys.argv) > 1 else 'generate'
    
    if action == 'generate':
        print(generate_license_key())
    
    elif action == 'validate':
        if len(sys.argv) < 3:
            print("Usage: python keygen.py validate <key>")
            sys.exit(1)
        
        result = validate_license_key(sys.argv[2])
        print(f"License: {sys.argv[2]}")
        if result['valid']:
            print(f"Status: ✓ VALID")
            print(f"Issued: {result['issued']}")
            print(f"Age: {result['age_days']} days")
        else:
            print(f"Status: ✗ INVALID")
            print(f"Error: {result['error']}")
    
    elif action == 'batch':
        count = int(sys.argv[2]) if len(sys.argv) > 2 else 10
        print(f"Generating {count} license keys:\n")
        for _ in range(count):
            print(generate_license_key())
            time.sleep(0.001)
    
    elif action == 'test':
        print("Testing key generation and validation...\n")
        
        key = generate_license_key()
        print(f"Generated: {key}")
        
        result = validate_license_key(key)
        print(f"Validation: {'✓ VALID' if result['valid'] else '✗ INVALID'}")
        if result['valid']:
            print(f"Issued: {result['issued']}")
        else:
            print(f"Error: {result['error']}")
        
        print("\nTesting invalid key:")
        invalid = 'AAAAA-BBBBB-CCCCC-DDDDD-EEEEE'
        result = validate_license_key(invalid)
        print(f"Key: {invalid}")
        print(f"Result: {'✗ INVALID' if not result['valid'] else '✓ VALID'}")
        if not result['valid']:
            print(f"Error: {result['error']}")
    
    else:
        print("Usage: python keygen.py [generate|validate|batch|test]")

if __name__ == '__main__':
    main()

File: keygen.php
<?php
/**
 * Ultra-secure license key generator for yurucode
 * Uses HMAC-SHA256 signatures with server-side secrets
 * No database required - pure cryptographic validation
 */

// Load server secrets
$envFile = __DIR__ . '/yurucode.com/.env.secret';
if (file_exists($envFile)) {
    $lines = file($envFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    foreach ($lines as $line) {
        if (strpos($line, '#') === 0) continue;
        if (strpos($line, '=') === false) continue;
        list($key, $value) = explode('=', $line, 2);
        putenv(trim($key) . '=' . trim($value));
    }
}

// Server secrets (loaded from environment)
define('MASTER_SECRET', getenv('YURUCODE_MASTER_SECRET') ?: '7f3a9c2e8b4d6f1a5c7e9b3d5f8a2c4e6b8d1f3a5c7e9b2d4f6a8c1e3b5d7f9a2c');
define('VALIDATION_SECRET', getenv('YURUCODE_VALIDATION_SECRET') ?: '9e2b4f7a1c5d8e3b6f9a2c5e8b1d4f7a3c6e9b2d5f8a1c4e7b9d3f6a8c2e5b7d');
define('TIMESTAMP_SECRET', getenv('YURUCODE_TIMESTAMP_SECRET') ?: '4a7c9e2b5d8f1a3c6e9b2d5f8a1c4e7b9d3f6a8c2e5b7d9f2a4c6e8b1d3f5a7c');

// License configuration
define('LICENSE_ALPHABET', '23456789ABCDEFGHJKLMNPQRSTUVWXYZ'); // 33 chars
define('LICENSE_VERSION', '2'); // Version 2 with HMAC

/**
 * Generate cryptographically secure random data
 */
function generate_random_bytes($length = 16) {
    return random_bytes($length);
}

/**
 * Encode bytes to base32-like custom alphabet
 * Encodes exactly 5 bytes to 8 characters
 */
function encode_to_base32($bytes) {
    // Pad to 5 bytes if needed
    $bytes = str_pad($bytes, 5, "\x00", STR_PAD_RIGHT);
    
    // Convert to binary string
    $binary = '';
    for ($i = 0; $i < 5; $i++) {
        $binary .= sprintf('%08b', ord($bytes[$i]));
    }
    
    // Split into 5-bit chunks (40 bits / 5 = 8 chunks)
    $result = '';
    for ($i = 0; $i < 8; $i++) {
        $chunk = substr($binary, $i * 5, 5);
        $index = bindec($chunk);
        $result .= LICENSE_ALPHABET[$index];
    }
    
    return $result;
}

/**
 * Decode from base32-like custom alphabet to bytes
 * Decodes 8 characters to exactly 5 bytes
 */
function decode_from_base32($str) {
    if (strlen($str) !== 8) return false;
    
    // Convert each character to 5-bit binary
    $binary = '';
    for ($i = 0; $i < 8; $i++) {
        $pos = strpos(LICENSE_ALPHABET, $str[$i]);
        if ($pos === false) return false;
        $binary .= sprintf('%05b', $pos);
    }
    
    // Convert binary string to bytes (40 bits = 5 bytes)
    $bytes = '';
    for ($i = 0; $i < 5; $i++) {
        $byte_binary = substr($binary, $i * 8, 8);
        $bytes .= chr(bindec($byte_binary));
    }
    
    return $bytes;
}

/**
 * Create HMAC signature for license data
 */
function create_license_signature($payload_segments, $timestamp) {
    // Combine all data for signing
    $data_to_sign = implode('|', [
        LICENSE_VERSION,
        implode('-', $payload_segments),
        $timestamp,
        VALIDATION_SECRET
    ]);
    
    // Create multi-layer HMAC
    $hmac1 = hash_hmac('sha256', $data_to_sign, MASTER_SECRET, true);
    $hmac2 = hash_hmac('sha256', $hmac1 . $timestamp, TIMESTAMP_SECRET, true);
    
    // Final signature combines both HMACs
    return hash('sha256', $hmac1 . $hmac2, true);
}

/**
 * Generate a new license key
 * Format: XXXXXXXX-XXXXXXXX-XXXXXXXX
 * Using base32-like encoding (8 chars = 5 bytes)
 * Total: 15 bytes (5+5+5)
 */
function generate_license_key() {
    // Generate random payload (6 bytes)
    $random_payload = generate_random_bytes(6);
    
    // Current timestamp (4 bytes, Unix timestamp)
    $timestamp_int = time();
    $timestamp = pack('N', $timestamp_int);
    
    // Combine payload and timestamp (10 bytes)
    $data = $random_payload . $timestamp;
    
    // Create signature
    $signature = create_license_signature([$random_payload, $timestamp], $timestamp_int);
    
    // Take first 5 bytes of signature
    $sig_short = substr($signature, 0, 5);
    
    // Encode three 5-byte segments
    $seg1 = encode_to_base32(substr($data, 0, 5));     // First 5 bytes of payload
    $seg2 = encode_to_base32(substr($data, 5, 5));     // Last byte of payload + timestamp
    $seg3 = encode_to_base32($sig_short);              // Signature
    
    // Return in familiar format (but with 8-char segments)
    return $seg1 . '-' . $seg2 . '-' . $seg3;
}

/**
 * Validate a license key (basic validation only)
 * Full validation must be done server-side
 */
function validate_license_format($key) {
    // Check format: three 8-character segments
    if (!preg_match('/^[' . LICENSE_ALPHABET . ']{8}(-[' . LICENSE_ALPHABET . ']{8}){2}$/', $key)) {
        // Also support old 5x5 format for compatibility
        if (!preg_match('/^[' . LICENSE_ALPHABET . ']{5}(-[' . LICENSE_ALPHABET . ']{5}){4}$/', $key)) {
            return false;
        }
    }
    
    return true;
}

/**
 * Validate license signature (server-side only)
 */
function validate_license_signature($key) {
    if (!validate_license_format($key)) {
        return ['valid' => false, 'error' => 'Invalid format'];
    }
    
    $segments = explode('-', strtoupper($key));
    
    // Handle new format (3 segments of 8 chars each)
    if (count($segments) === 3 && strlen($segments[0]) === 8) {
        // Decode segments
        $seg1_bytes = decode_from_base32($segments[0]);
        $seg2_bytes = decode_from_base32($segments[1]);
        $seg3_bytes = decode_from_base32($segments[2]);
        
        if ($seg1_bytes === false || $seg2_bytes === false || $seg3_bytes === false) {
            return ['valid' => false, 'error' => 'Decode failed'];
        }
        
        // Extract data
        $payload = $seg1_bytes . substr($seg2_bytes, 0, 1); // 6 bytes
        $timestamp_bytes = substr($seg2_bytes, 1, 4);       // 4 bytes
        $provided_signature = $seg3_bytes;                   // 5 bytes
        
        if (strlen($timestamp_bytes) !== 4) {
            return ['valid' => false, 'error' => 'Invalid timestamp'];
        }
        
        $timestamp = unpack('N', $timestamp_bytes)[1];
        
        // Check timestamp validity
        $current_time = time();
        if ($timestamp > $current_time + 86400) {
            return ['valid' => false, 'error' => 'Timestamp in future'];
        }
        if ($timestamp < $current_time - (365 * 86400 * 2)) {
            return ['valid' => false, 'error' => 'License too old'];
        }
        
        // Recreate signature
        $expected_signature = create_license_signature([$payload, $timestamp_bytes], $timestamp);
        
        // Compare signatures (first 5 bytes)
        if (substr($expected_signature, 0, 5) !== $provided_signature) {
            return ['valid' => false, 'error' => 'Invalid signature'];
        }
        
        return [
            'valid' => true,
            'timestamp' => $timestamp,
            'issued' => date('Y-m-d H:i:s', $timestamp),
            'age_days' => floor(($current_time - $timestamp) / 86400)
        ];
    }
    
    // Old format not supported with new validation
    return ['valid' => false, 'error' => 'Unsupported key format'];
}

/**
 * Batch generate licenses
 */
function generate_batch($count = 10) {
    $keys = [];
    for ($i = 0; $i < $count; $i++) {
        $keys[] = generate_license_key();
        usleep(1000); // Small delay to ensure timestamp variation
    }
    return $keys;
}

// CLI interface
if (php_sapi_name() === 'cli') {
    $argv = $_SERVER['argv'];
    $argc = $_SERVER['argc'];
    
    if ($argc > 1) {
        switch ($argv[1]) {
            case '--validate':
                if ($argc > 2) {
                    $key = $argv[2];
                    $result = validate_license_signature($key);
                    echo "License: $key\n";
                    if ($result['valid']) {
                        echo "Status: ✓ VALID\n";
                        echo "Issued: {$result['issued']}\n";
                        echo "Age: {$result['age_days']} days\n";
                    } else {
                        echo "Status: ✗ INVALID\n";
                        echo "Error: {$result['error']}\n";
                    }
                } else {
                    echo "Usage: php keygen.php --validate <key>\n";
                }
                break;
                
            case '--batch':
                $count = ($argc > 2) ? intval($argv[2]) : 10;
                echo "Generating $count license keys:\n\n";
                $keys = generate_batch($count);
                foreach ($keys as $key) {
                    echo $key . "\n";
                }
                break;
                
            case '--test':
                echo "Testing key generation and validation...\n\n";
                
                // Generate a key
                $key = generate_license_key();
                echo "Generated: $key\n";
                
                // Validate it
                $result = validate_license_signature($key);
                echo "Validation: " . ($result['valid'] ? '✓ VALID' : '✗ INVALID') . "\n";
                if ($result['valid']) {
                    echo "Timestamp: {$result['issued']}\n";
                }
                
                // Test invalid key
                echo "\nTesting invalid key:\n";
                $invalid = 'AAAAA-BBBBB-CCCCC-DDDDD-EEEEE';
                $result = validate_license_signature($invalid);
                echo "Key: $invalid\n";
                echo "Result: " . ($result['valid'] ? '✓ VALID' : '✗ INVALID') . "\n";
                if (!$result['valid']) {
                    echo "Error: {$result['error']}\n";
                }
                
                // Test modified key
                echo "\nTesting modified key:\n";
                $parts = explode('-', $key);
                $parts[4] = 'ZZZZZ'; // Modify signature
                $modified = implode('-', $parts);
                $result = validate_license_signature($modified);
                echo "Modified: $modified\n";
                echo "Result: " . ($result['valid'] ? '✓ VALID' : '✗ INVALID') . "\n";
                if (!$result['valid']) {
                    echo "Error: {$result['error']}\n";
                }
                break;
                
            default:
                echo "yurucode License Generator v2.0\n";
                echo "Usage:\n";
                echo "  php keygen.php                    Generate single key\n";
                echo "  php keygen.php --validate <key>   Validate a key\n";
                echo "  php keygen.php --batch [count]    Generate multiple keys\n";
                echo "  php keygen.php --test             Run tests\n";
                break;
        }
    } else {
        // Default: generate single key
        echo generate_license_key() . "\n";
    }
}